apps/emqx_resource/src/emqx_resource_manager.erl

```diff
@@ -2100,3 +2223,76 @@ abort_health_checks_for_channel(Data0, ChannelId) ->
         hc_workers = HCWorkers,
         hc_pending_callers = Pending
     }.
+
+collect_and_handle_channel_operations_connected(Op, Data0) ->
+    From = undefined,
+    Ops = collect_and_compress_channel_operations([Op]),
+    lists:foldl(
+        fun
+            (#add_channel{channel_id = ChannelId, config = Config}, {AccActions, AccData}) ->
+                {Actions, Data} = handle_add_channel(From, AccData, ChannelId, Config),
+                {Actions ++ AccActions, Data};
+            (#remove_channel{channel_id = ChannelId}, {AccActions, AccData}) ->
+                {Actions, Data} = handle_remove_channel(From, ChannelId, AccData),
+                {Actions ++ AccActions, Data}
+        end,
+        {[], Data0},
+        Ops
+    ).
+
+collect_and_handle_channel_operations_not_connected(Op, State, Data0) ->
+    From = undefined,
+    Ops = collect_and_compress_channel_operations([Op]),
+    lists:foldl(
+        fun
+            (#add_channel{channel_id = ChannelId, config = Config}, {AccActions, AccData}) ->
+                {Actions, Data} =
+                    handle_not_connected_add_channel(From, ChannelId, Config, State, AccData),
+                {Actions ++ AccActions, Data};
+            (#remove_channel{channel_id = ChannelId}, {AccActions, AccData}) ->
+                {Actions, Data} = handle_remove_channel(From, ChannelId, AccData),
+                {Actions ++ AccActions, Data}
+        end,
+        {[], Data0},
+        Ops
+    ).
+
+collect_and_compress_channel_operations(Acc) ->
+    MaxRequests = 500,
+    Ops0 = collect_channel_operations(MaxRequests, Acc),
+    {Ops1, _} = lists:foldl(
+        fun(Op, {OpAcc, N}) ->
+            ChannelId = operation_channel_id(Op),
+            case OpAcc of
+                #{ChannelId := #{op := _} = Old} ->
+                    %% Just replace with newer
+                    {OpAcc#{ChannelId := Old#{op := Op}}, N};
+                #{} ->
+                    %% New channel id
+                    New = #{op => Op, n => N},
+                    {OpAcc#{ChannelId => New}, N + 1}
+            end
+        end,
+        {#{}, 0},
+        Ops0
+    ),
+    Ops2 = lists:map(fun(#{op := Op, n := N}) -> {N, Op} end, maps:values(Ops1)),
+    Ops = lists:keysort(1, Ops2),
+    lists:map(fun({_N, Op}) -> Op end, Ops).
+
+operation_channel_id(#add_channel{channel_id = ChannelId, config = _Config}) ->
+    ChannelId;
+operation_channel_id(#remove_channel{channel_id = ChannelId}) ->
+    ChannelId.
+
+collect_channel_operations(0, Acc) ->
```

Comment:
nit: collect and compress can be done in one go?
use `lists:keyfind` to find the channel ID to preserve the order.
so there is no need for indexing map to list, sorting etc.

Reply:
not sure I understand, but that requires traversing the accumulator multiple times searching for the channel id, right?

I think we won't have so many messages that this is a performance issue.